### PR TITLE
zeroex: Validate coordinator orders

### PR DESCRIPTION
**Pull request by fabioberger**
Fixes: #100

### CODE REVIEW

File: zeroex/order_validator_test.go [hunk 1]
```diff
@@ -151,6 +158,79 @@ func TestBatchValidateSignatureInvalid(t *testing.T) {
...
+
+	ethClient, err := ethclient.Dial(constants.GanacheEndpoint)
+	require.NoError(t, err)
+
+	orderValidator, err := NewOrderValidator(ethClient, constants.TestNetworkID)
+	require.NoError(t, err)
+
+	validationResults := orderValidator.BatchValidate(signedOrders)
+	assert.Len(t, validationResults.Accepted, 0)
+	assert.Len(t, validationResults.Rejected, 1)
```
Comment (albrow) [inline zeroex/order_validator_test.go]: Same comment here regarding panics when `validationResults.Rejected` is empty.

File: core/message_handler.go [hunk 2]
```diff
@@ -144,7 +144,7 @@ func (app *App) HandleMessages(messages []*p2p.Message) error {
 			"from":              msg.From.String(),
 		}).Trace("not storing rejected order received from peer")
 		switch rejectedOrderInfo.Status {
-		case ROInternalError, ROOrderAlreadyStored, zeroex.RORequestFailed:
+		case ROInternalError, ROOrderAlreadyStored, zeroex.ROEthRPCRequestFailed:
```
Comment (albrow) [inline core/message_handler.go]: I think we should also add `ROCoordinatorRequestFailed` to the list of errors that should not incur a negative score?

File: zeroex/order_validator.go [hunk 3]
```diff
@@ -322,6 +359,161 @@ func (o *OrderValidator) BatchValidate(rawSignedOrders []*SignedOrder) *Validati
...
+			continue
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if resp.StatusCode != 200 {
+			log.WithFields(map[string]interface{}{
+				"endpoint":   endpoint,
+				"statusCode": resp.StatusCode,
+				"requstURL":  requestURL,
+				"body":       string(body),
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply zeroex/order_validator.go]: If we got a non-200 status code, I'd expect the body to contain an error object, but not a very large one. I think it's worth logging.

File: zeroex/order_validator.go [hunk 4]
```diff
@@ -322,6 +359,161 @@ func (o *OrderValidator) BatchValidate(rawSignedOrders []*SignedOrder) *Validati
...
+					OrderHash:   orderHash,
+					SignedOrder: signedOrder,
+					Kind:        MeshError,
+					Status:      ROCoordinatorRequestFailed,
+				})
+			}
+			continue
+		}
+		var response softCancelResponse
+		err = json.Unmarshal(body, &response)
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply zeroex/order_validator.go]: We need the body as a string above anyways so I'm going to keep using `Unmarshal` here.

File: zeroex/order_validator.go [hunk 5]
```diff
@@ -363,6 +363,12 @@ type softCancelResponse struct {
 	OrderHashes []common.Hash `json:"orderHashes"`
 }
 
+/**
+ * batchValidateSoftCancelled validates any order specifying the Coordinator contract as the `senderAddress` to ensure
```
Comment (albrow) [inline zeroex/order_validator.go]: This style of comment isn't idiomatic and probably won't render correctly on GoDoc or in editors which support inline documentation on hover. Can you change it to:

```go
// batchValidateSoftCancelled validates any order specifying the Coordinator contract as the `senderAddress` to ensure
// that it hasn't been cancelled off-chain (soft cancellation). It does this by looking up the Coordinator server endpoint
// ...
```

Or if you still want to use multi-line comments, remove the extra asterisks:

```go
/*
batchValidateSoftCancelled validates any order specifying the Coordinator contract as the `senderAddress` to ensure
that it hasn't been cancelled off-chain (soft cancellation). It does this by looking up the Coordinator server endpoint
...
*/
```

File: zeroex/order_validator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (albrow) [reply zeroex/order_validator.go]: It looks like we only need to read the body as a string if there's an error. We could still make the code more efficient by using a single buffer for the non-error case. But I'm also okay with it if you think the version we have now is cleaner.

**Pull request merged**